### PR TITLE
Implement to_dict and to_pandas for Dataset

### PR DESCRIPTION
**Pull request by SBrandeis**
With options to return a generator or the full dataset

### CODE REVIEW

**Review by lhoestq — COMMENTED** [review]
Nice thanks !

Also can you use `query_table` with self._indices (same change as I suggested in the to_csv PR) ? This way this will support shuffled/sharded datasets

File: src/datasets/arrow_dataset.py [hunk 1]
```diff
@@ -2181,6 +2182,58 @@ def generator():
...
+            return self._data.to_pydict()
+        else:
+            batch_size = batch_size if batch_size else config.DEFAULT_MAX_BATCH_SIZE
+            return (
+                query_table(
+                    pa_table=self._data,
+                    key=slice(offset, offset + batch_size),
+                    indices=self._indices.column(0) if self._indices is not None else None,
+                ).to_pydict()
+                for offset in range(0, self._data.num_rows, batch_size)
```
Comment (lhoestq) [inline src/datasets/arrow_dataset.py:2207]:
```suggestion
                for offset in range(0, len(self), batch_size)
```
if the dataset has an indices mapping, then probably the size of the dataset is not the size of the underlying table, but actually the size of the indices mapping. 
So you can use `len(self)` here, that returns the size of the dataset in both cases.

File: src/datasets/arrow_dataset.py [hunk 2]
```diff
@@ -2181,6 +2182,58 @@ def generator():
...
+            batched (`bool`): Set to :obj:`True` to return a generator that yields the dataset as batches of
+                :param:`batch_size` rows. Defaults to :obj:`False` (returns the whole datasetas once)
+            bacth_size (`Optional[int]`): The size (number of rows) of the batches if :param:`batched` is `True`.
+                Defaults to :obj:`.config.DEFAULT_MAX_BATCH_SIZE`.
+
+        Returns:
+            `dict` or `Iterator[dict]`
+        """
+        if not batched:
+            return self._data.to_pydict()
```
Comment (lhoestq) [inline src/datasets/arrow_dataset.py:2198]: You will need to use query_table here as well to take the indices mapping into account (`indices=self._indices.column(0) if self._indices is not None else None`).

Feel free to use the key `slice(0, len(self))` to query the full table after applying the indices mapping.

File: src/datasets/arrow_dataset.py [hunk 3]
```diff
@@ -2181,6 +2182,58 @@ def generator():
...
+            return self._data.to_pandas()
+        else:
+            batch_size = batch_size if batch_size else config.DEFAULT_MAX_BATCH_SIZE
+            return (
+                query_table(
+                    pa_table=self._data,
+                    key=slice(offset, offset + batch_size),
+                    indices=self._indices.column(0) if self._indices is not None else None,
+                ).to_pandas()
+                for offset in range(0, self._data.num_rows, batch_size)
```
Comment (lhoestq) [inline src/datasets/arrow_dataset.py:2234]: same: 
```suggestion
                for offset in range(0, len(self), batch_size)
```

File: src/datasets/arrow_dataset.py [hunk 4]
```diff
@@ -2181,6 +2182,58 @@ def generator():
...
+            batched (`bool`): Set to :obj:`True` to return a generator that yields the dataset as batches of
+                :param:`batch_size` rows. Defaults to :obj:`False` (returns the whole datasetas once)
+            bacth_size (`Optional[int]`): The size (number of rows) of the batches if :param:`batched` is `True`.
+                Defaults to :obj:`.config.DEFAULT_MAX_BATCH_SIZE`.
+
+        Returns:
+            `pandas.DataFrame` or `Iterator[pandas.DataFrame]`
+        """
+        if not batched:
+            return self._data.to_pandas()
```
Comment (lhoestq) [inline src/datasets/arrow_dataset.py:2225]: same: use query_table

File: tests/test_arrow_dataset.py [hunk 5]
```diff
@@ -1222,6 +1222,52 @@ def test_export(self, in_memory):
...
+                    self.assertLessEqual(len(batch[col_name]), bacth_size)
+
+            # Full
+            dset_to_dict = dset.to_dict()
+            self.assertIsInstance(dset_to_dict, dict)
+            self.assertListEqual(sorted(dset_to_dict.keys()), sorted(dset.column_names))
+            for col_name in dset.column_names:
+                self.assertIsInstance(dset_to_dict[col_name], list)
+                self.assertLessEqual(len(dset_to_dict[col_name]), dset.num_rows)
+
```
Comment (lhoestq) [inline tests/test_arrow_dataset.py:1246]: feel free to add a test where there is an indices mapping
```
dset_with_indices_mapping = dset.select([1, 0])
```
for example.
The new `dset_with_indices_mapping` dataset is of size 2 and its first element is at the position 1 in the pyarrow table, and the second element is at the position 0 in the pyrrow table.

**Review by lhoestq — COMMENTED** [review]
Nice ! Thanks for adding the support for datasets with an indices mapping.
I added some more comments about query_table and the batched=False case, as well as suggestions to use `len(self)` instead of `self._data.num_rows` in query_table

File: src/datasets/arrow_dataset.py (hunk 1 above)
Comment (SBrandeis) [reply src/datasets/arrow_dataset.py:2207]: Woops

**Review by lhoestq — APPROVED** [review]
Nice thank you !

**Comment by lhoestq** [comment]
Next step is going to add these two in the documentation ^^

**Pull request merged**